### PR TITLE
subjects: create subjects specific vocabulary

### LINKED ISSUE #63: Split Subjects into its own Vocabulary
Review list of steps to do like affiliations but for subjects:

**Data layer**

- [x] Define relation in CommonFIeldsMixin.
    - [x] Determine which fields to dump in ES index.
- [x] JSONShema: Do not modify an already released schema, create new version instead.
- [x] JSONSchema: Define or review existing property
    - List or single?
    - Mixed linked with non-linked?
- [x] Mappings adapt to JSONSchema and dumped fields.

**Service layer**

- [x] Schema: Fix metadata schema.
- [x] Facets: Do we need facets? Guillaume: let's skip for now

**Presentation layer**

- [x] Fix REST API serializations
- [x] Fix other serialization formats.

**Tests**

- [x] Add tests that record linking is wokring as expected (e.g. from service layer).
    - Bad and good data.
- [x] Add REST API tests for serialization, facets
- [x] Ensure that I18N is tested properly if required.

Follow structure from:

https://codimd.web.cern.ch/vc6wJipAS66l1XjwKjYdmA?both#

### PR DESCRIPTION
**Pull request by fenekku**
closes #63

### CODE REVIEW

File: invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json:11]: there is a definition `identifier`, which in the end is a string. We can use it here.

File: invenio_vocabularies/contrib/subjects/record_type.py
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/record_type.py:1]: I think we are calling this file `Affiliations.py`, so either we call this `subjects.py` or we rename affiliations. I'm leaning towards `subjects.py`

File: invenio_vocabularies/contrib/subjects/schema.py
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/schema.py:17]: even if not all the fields are used, we should inherit from `BaseVocabularySchema`. Mainly to keep consistency with the RFC 🤷‍♂️

**Comment by ppanero** [comment]
very tiny comments, a really nice work @fenekku. I'm taking it from here, the only thing needed is an agreement on the actual data model of it. Hopefully with @lnielsen we can set it this morning and merge this before your shift 🚀 

I'm pushing changes in a different commit just to make easier to spot changes, but we should squash before merging hehe

File: invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json:13]: data mode is tbd, but should we call this `subject`? UI is ready to configure this field name. Name makes sense too

File: tests/contrib/affiliations/test_affiliations_jsonschema.py
Comment (ppanero) [inline tests/contrib/affiliations/test_affiliations_jsonschema.py:85]: he, nice catch

File: tests/contrib/subjects/test_subjects_jsonschema.py
Comment (ppanero) [inline tests/contrib/subjects/test_subjects_jsonschema.py:39]: can be both... even "should" be both. As far as I understood `id` is the PK in the end, but `name` is the label shown to the user (at least in affiliations).

File: tests/contrib/subjects/test_subjects_jsonschema.py
Comment (ppanero) [inline tests/contrib/subjects/test_subjects_jsonschema.py:48]: Do we forsee it in others? @lnielsen otherwise I'd remove and just keep name

File: tests/contrib/subjects/test_subjects_service.py
Comment (ppanero) [inline tests/contrib/subjects/test_subjects_service.py:18]: I think we have a difference on use cases... you are pointing towards "users can add subjects to a subtype" and for affiliations what we did was "free text/custom ones are stored in the record". That means:
- No aggregations possible
- No sharing between records

**Comment by ppanero** [comment]
Subjects have been implemented in the same manner as affiliations. Things to discuss:
- Data model
- Use case of custom subjects being sabed to the db instead of just attach them to a record

File: invenio_vocabularies/contrib/subjects/record_type.py
Comment (ppanero) [reply invenio_vocabularies/contrib/subjects/record_type.py]: Saw the pre-existing `subjects.py` I guess it was done like this to avoid removing the old one. I move it to `subjects.py`.

**Review by fenekku — COMMENTED** [review]
Better late than never?

File: invenio_vocabularies/contrib/subjects/api.py
Comment (fenekku) [inline invenio_vocabularies/contrib/subjects/api.py:14]: I originally made this one singular because `Record` in api.py is singular too.

File: invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json
Comment (fenekku) [inline invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json:16]: Like was brought up, I am not sure anymore if we even want to store custom keywords here actually. Just storing them on the record is fine.

**Comment by fenekku** [comment]
@ppanero I squashed everything into 1 commit, so git reset --hard on this branch locally if you need to modify it further (but you shouldn't have to). 

In a separate PR I will remove filtering on generic vocabularies (and related things), but this one is good to be reviewed for separate subject. I will look into loading tomorrow and adapt Northwestern's invenio-subjects-mesh to this new structure.

File: invenio_vocabularies/cli.py
Comment (ppanero) [inline invenio_vocabularies/cli.py:21]: this import breaks below, it's using the old `subject_record_type`

File: invenio_vocabularies/contrib/subjects/config.py
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/config.py:27]: This works for `search_as_you_type` fields. I've added that to v7 subject (not available in v6)

File: invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/jsonschemas/subjects/subject-v1.0.0.json:16]: yes, for the PIDField to be "re-built" easily if I recall correctly.

File: invenio_vocabularies/contrib/subjects/resources.py
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/resources.py:15]: 🚀

File: invenio_vocabularies/contrib/subjects/schema.py
Comment (ppanero) [reply invenio_vocabularies/contrib/subjects/schema.py:17]: This still applies. It was agreed we should comply with the RFC when doing affiliations. We can discuss this with Lars on Monday, but I would go for the RFC since is what has been agree for the rest. If we decide to change, we need to change Affiliations too.

File: invenio_vocabularies/contrib/subjects/schema.py
Comment (ppanero) [inline invenio_vocabularies/contrib/subjects/schema.py:20]: agree on this.

File: run-tests.sh
Comment (ppanero) [reply run-tests.sh:23]: I'm not using pipenv and it's not happening to me 😅 I think we should leave it generic (or add it everywhere, like every module).

File: tests/contrib/subjects/test_subjects_api.py
Comment (ppanero) [inline tests/contrib/subjects/test_subjects_api.py:56]: 😅

File: tests/contrib/subjects/test_subjects_resource.py
Comment (ppanero) [inline tests/contrib/subjects/test_subjects_resource.py:127]: why the `, res.json`?

File: tests/contrib/subjects/test_subjects_resource.py
Comment (ppanero) [inline tests/contrib/subjects/test_subjects_resource.py:119]: We should try to test order here, but with this is pretty complex and it is not as easy since is a filtered suggestion...

**Review by ppanero — APPROVED** [review]
some minor changes (e.g. one import fix)

**Comment by ppanero** [comment]
I am merging and releasing to unblock things like the frontend. If any points need revisting we can discuss later on/monday and fix in a different PR.

File: tests/contrib/subjects/test_subjects_resource.py
Comment (fenekku) [reply tests/contrib/subjects/test_subjects_resource.py]: So that on a failed assert, the json is shown. I've found this to be much more efficient in testing/debugging as it directly shows the state of things.

File: invenio_vocabularies/cli.py
Comment (fenekku) [inline invenio_vocabularies/cli.py:22]: Yeah this file needs to be changed. The create subjects vocabulary below is new to me and certainly doesn't work now. I will still do the loading in rdm-records, but I wouldn't mind eventually moving it back here if it makes sense.

**Pull request merged**
